### ISSUE TITLE
question about margin_loss

### ISSUE THREAD

**Issue by jingzhengli** [snap-stanford/orca#5]
Hi, thanks for sharing the nice code!
I have some questions about "margin loss " as follows: the code of margin loss is the same as cross_entropy_loss. I find the value of "output" is the same as the value of "x". what does the code of margin loss do?

_class MarginLoss(nn.Module):
    
    def __init__(self, m=0.2, weight=None, s=10):
        super(MarginLoss, self).__init__()
        self.m = m
        self.s = s
        self.weight = weight

    def forward(self, **x**, target):
        index = torch.zeros_like(x, dtype=torch.uint8)
        index.scatter_(1, target.data.view(-1, 1), 1)
        x_m = x - self.m * self.s
    
        output = torch.where(index, x_m, x)
        return F.cross_entropy(**output**, target, weight=self.weight)_

**Comment by jingzhengli** [comment]
> Hi, thanks for sharing the nice code! I have some questions about "margin loss " as follows: the code of margin loss is the same as cross_entropy_loss. I find the value of "output" is the same as the value of "x". what does the code of margin loss do?
> 
> _class MarginLoss(nn.Module):
> 
> ```
> def __init__(self, m=0.2, weight=None, s=10):
>     super(MarginLoss, self).__init__()
>     self.m = m
>     self.s = s
>     self.weight = weight
> 
> def forward(self, **x**, target):
>     index = torch.zeros_like(x, dtype=torch.uint8)
>     index.scatter_(1, target.data.view(-1, 1), 1)
>     x_m = x - self.m * self.s
> 
>     output = torch.where(index, x_m, x)
>     return F.cross_entropy(**output**, target, weight=self.weight)_
> ```

sorry, I misunderstood. Nice code

**Comment by Feobi1999** [comment]
@jingzhengli 
Could you please tell me how to match this loss implementation with Equation 4 in the paper? I don't see the part about calculating uncertainty, can we subtract a fixed value from the unlabeled output to get that?
Thanks!

**Comment by July-blue** [comment]
> 嗨，感谢您分享漂亮的代码！我对“保证金损失”有一些问题，如下所示：保证金损失的代码与cross_entropy_loss相同。我发现“输出”的值与“x”的值相同。保证金损失的代码是做什么的？
> 
> _class 保证金损失（nn.模块）：
> 
> ```
> def __init__(self, m=0.2, weight=None, s=10):
>     super(MarginLoss, self).__init__()
>     self.m = m
>     self.s = s
>     self.weight = weight
> 
> def forward(self, **x**, target):
>     index = torch.zeros_like(x, dtype=torch.uint8)
>     index.scatter_(1, target.data.view(-1, 1), 1)
>     x_m = x - self.m * self.s
> 
>     output = torch.where(index, x_m, x)
>     return F.cross_entropy(**output**, target, weight=self.weight)_
> ```



> Hi, thanks for sharing the nice code! I have some questions about "margin loss " as follows: the code of margin loss is the same as cross_entropy_loss. I find the value of "output" is the same as the value of "x". what does the code of margin loss do?
> 
> _class MarginLoss(nn.Module):
> 
> ```
> def __init__(self, m=0.2, weight=None, s=10):
>     super(MarginLoss, self).__init__()
>     self.m = m
>     self.s = s
>     self.weight = weight
> 
> def forward(self, **x**, target):
>     index = torch.zeros_like(x, dtype=torch.uint8)
>     index.scatter_(1, target.data.view(-1, 1), 1)
>     x_m = x - self.m * self.s
> 
>     output = torch.where(index, x_m, x)
>     return F.cross_entropy(**output**, target, weight=self.weight)_
> ```

@jingzhengli Hello, I have the same puzzle as you, but I still don't understand it. Could you please share your understanding? Thank you!!

**Comment by jingzhengli** [comment]
> > 嗨，感谢您分享漂亮的代码！我对“保证金损失”有一些问题，如下所示：保证金损失的代码与cross_entropy_loss相同。我发现“输出”的值与“x”的值相同。保证金损失的代码是做什么的？
> > _class 保证金损失（nn.模块）：
> > ```
> > def __init__(self, m=0.2, weight=None, s=10):
> >     super(MarginLoss, self).__init__()
> >     self.m = m
> >     self.s = s
> >     self.weight = weight
> > 
> > def forward(self, **x**, target):
> >     index = torch.zeros_like(x, dtype=torch.uint8)
> >     index.scatter_(1, target.data.view(-1, 1), 1)
> >     x_m = x - self.m * self.s
> > 
> >     output = torch.where(index, x_m, x)
> >     return F.cross_entropy(**output**, target, weight=self.weight)_
> > ```
> 
> > Hi, thanks for sharing the nice code! I have some questions about "margin loss " as follows: the code of margin loss is the same as cross_entropy_loss. I find the value of "output" is the same as the value of "x". what does the code of margin loss do?
> > _class MarginLoss(nn.Module):
> > ```
> > def __init__(self, m=0.2, weight=None, s=10):
> >     super(MarginLoss, self).__init__()
> >     self.m = m
> >     self.s = s
> >     self.weight = weight
> > 
> > def forward(self, **x**, target):
> >     index = torch.zeros_like(x, dtype=torch.uint8)
> >     index.scatter_(1, target.data.view(-1, 1), 1)
> >     x_m = x - self.m * self.s
> > 
> >     output = torch.where(index, x_m, x)
> >     return F.cross_entropy(**output**, target, weight=self.weight)_
> > ```
> 
> @jingzhengli Hello, I have the same puzzle as you, but I still don't understand it. Could you please share your understanding? Thank you!!

抱歉这么晚回复。这个代码是这样的，首先 ，“x_m = x - self.m * self.s” 计算的是模型输出的原始的logits减去margin的结果。然后，你不理解的原因可能是没看懂这一行：“output = torch.where(index, x_m, x)”，这一步把原始的模型输出的logits，使用减去margin的logits按照index替换掉了。所以最后outpu就是减去margin的模型输出，再计算的正常的交叉上损失就行。希望对你有帮助。

**Comment by OverfitFlow** [comment]
Hi Jingzheng,

it seems there is also no scaling for logits here, which is related to the temperature softmax.
